Add graph documentations (#788)
marian-nmt/marian-dev: doc/graph.md

### Forward pass

The forward pass refers to the calculation process. 
It traverses through all nodes from input layer (leaves) to output layer (root). 
To perform the forward pass, you can call function `forward()`. The `forward()` function mainly does two things:

- allocates memory for each node (`Node::allocate()`)
- computing the new tensor for each node by performing required operations (`Node::forward()`), and the resulting new tensor is stored in `val_` attribute in each Node.

**Forward pass of Example 1**

To run the forward pass of Example 1, you can run the following code:

```cpp
// Perform the forward pass on the nodes of the graph
graph->forward();
// get the computation result of z
std::vector<float> w;
z->val()->get(w);
std::cout<<"z="<<w[0]<<std::endl;
// The output is: z=6.9093
```

### **Backward pass**

The backward pass refers to the process of computing the output error. 
It traverses through all *trainable* nodes from output layer to input layer. 
You can call `backward()` to perform the backward pass. 
The `backward()` function mainly computes the gradients using the chain rule:

- allocates memory and initialise gradients for each *trainable* Node

> - computes the gradients based on backward steps (`Node::backwardOps()`) from each Node, and accumulates them in `adj_` attribute in each Node

Author: I changed accumulates to stores to remove unclearness.

   graph->backward();
*/
```

**Backward pass of modified Example 1**

As shown in Figure 2, there is no trainable node in the graph of Example 1; 
this means we cannot perform backward pass on this graph. 
To demonstrate the backward pass, we modify Example 1 by changing constant node `x` to a parameter node (change `constant()` to `param()`). 
Here is the modification:

```cpp
// add parameter node x
auto x = graph->param("x", {1,1}, inits::fromValue(2));
```

The resulting graph is also different as displayed in Figure 3.

![fig3](images/example1_dot2.png "Figure 3 Graph layout of modified Example 1")

*Figure 3 Graph layout of modified Example 1*

To perform the backward pass of modified Example 1, you can run the following code:

```cpp
// Perform the backward pass on the trainable nodes of the graph
graph->backward();
// get the gradient of x node
std::vector<float> b;
x->grad()->get(b);
std::cout<<"dz/dx="<<b[0]<<std::endl;
// The output is: dz/dx=2.58385
```

### Optimiser

After the backward pass, we obtain the gradients of the leaves. 
However, the job is not done yet. 
To train a model, we need to update the model parameters according to the gradients.

> This comes to how to define the loss function and optimiser for the graph. 

Reviewer: ```suggestion
This comes from how we define the loss function and optimiser for the graph. 
```

A loss function is used to calculate the model error between the predicated value and the actual value. 
The goal is to minimise this error during training. 
In a graph, the loss function is also represented as a group of node(s). 
You can also use the operators provided in [`expression_operators.h`](api/file_src_graph_expression_operators.h.html#file-src-graph-expression-operators-h) file to define the loss function. 
E.g., Marian offers `cross_entropy()` function to compute the cross-entropy loss between true labels and predicted labels.

**Define a loss function for modified Example 1**

Suppose we know the actual value of `z` is 6 with `y = 3`, and `x` is the parameter we would like to learn from the model. 
The loss function we choose here is the absolute error:

```cpp
// pass the actual value to the model
auto actual = graph->constant({1,1}, inits::fromValue(6));
// define loss function
auto loss = abs(actual-z);
```

The graph is changed to Figure 4. 

![fig4](images/example1_dot3.png "Figure 4 Graph layout of modified Example 1 with loss function")

*Figure 4 Graph layout of modified Example 1 with loss function*

> The optimiser is to adjust the variables to fit the data. 

Reviewer: ```suggestion
The purpose of the optimiser is to adjust the variables to fit the data. 
```

In Marian, there are three built-in optimiser classes: `Sgd`, `Adagrad` and `Adam`.
`Sgd` is an optimiser based on [stochastic gradient descent](https://en.wikipedia.org/wiki/Stochastic_gradient_descent). 
For each iteration, it updates the parameter `w` according the rule of `w = w - learning_rate * gradient`. 

> `Adagrad` implements [Adagrad algorithm](https://www.jmlr.org/papers/volume12/duchi11a/duchi11a.pdf), an optimiser with parameter-specific learning rates, which are adapted relative to how frequently a parameter gets updated during training.

Reviewer: ```suggestion
`Adagrad` implements the [Adagrad algorithm](https://www.jmlr.org/papers/volume12/duchi11a/duchi11a.pdf), an optimiser with parameter-specific learning rates, which are adapted relative to how frequently a parameter gets updated during training.
```

> `Adam` is based on [Adam algorithm](https://arxiv.org/abs/1412.6980), a stochastic gradient descent method that is based on adaptive estimation of first-order and second-order moments. 

Reviewer: ```suggestion
`Adam` is an implementation of the [Adam algorithm](https://arxiv.org/abs/1412.6980), a stochastic gradient descent method that is based on an adaptive estimation of first-order and second-order moments. 
```

**Debugging for modified Example 1**

Suppose we want to check the results of node `x` during the computation. We can call `debug()` to mark node `x` for debugging. 
```cpp
// mark node x for debugging with logging message "Parameter x"
debug(x, "Parameter x");
```
The output is shown as follows with `createLoggers()`:
```cpp
[2021-02-16 15:10:51] [memory] Reserving 256 B, device gpu0
[2021-02-16 15:10:51] Debug: Parameter x op=param
[2021-02-16 15:10:51] shape=1x1 size=1 type=float32 device=gpu0 ptr=140505547538432 bytes=256
min: 2.00000000 max: 2.00000000 l2-norm: 2.00000000
[[   2.00000000 ]]

[2021-02-16 15:10:51] [memory] Reserving 256 B, device gpu0
[2021-02-16 15:10:51] Debug Grad: Parameter x op=param
[2021-02-16 15:10:51] shape=1x1 size=1 type=float32 device=gpu0 ptr=140505547538944 bytes=256
min: 2.58385324 max: 2.58385324 l2-norm: 2.58385324
[[   2.58385324 ]]
```

### More advanced

For more details about graph execution, a graph keeps track of all the `Node` objects in its `nodesForward_` and `nodesBackward_` lists. 
`nodesForward_` contains all nodes used for the forward pass and `nodesBackward_` contains all trainable nodes used for the backward pass. 

> All the tensor objects for a graph is stored in its `tensors_` attribute. 

Reviewer: ```suggestion
All the tensor objects for a graph are stored in its `tensors_` attribute. 
```

`tensors_` is a shared pointer holding memory and nodes for a graph. 
Since each `Node` can result in new tensors, this attribute is used to allocate memory for new tensors during the forward and backward pass.
This `tensors_` attribute gets cleared before a new graph is built. 
Another important attribute in `ExpressionGraph` is `paramsByElementType_`. 
This attribute holds memory and nodes that corresponds to graph parameters.
You can call `params()` function in a graph to get all the parameter objects:

```cpp
// return the Parameters object related to the graph
// The Parameters object holds the whole set of the parameter nodes.
graph->params();
```

Besides, we provide APIs to support the mechanism of Gradient Checkpointing. 
This method works by trading compute for memory, which reruns a forward-pass segment for each checkpoint segment during backward pass. 

> Currently, Marian only support set checkpoint nodes manually (by calling `Node::markCheckpoint()` or `checkpoint()`). 

Reviewer: ```suggestion
Currently, Marian only supports setting checkpoint nodes manually by calling `Node::markCheckpoint()` or `checkpoint()`. 
```

> To enable the mode of gradient-checkpointing for a graph, we use `setCheckpointing()`:

Reviewer: ```suggestion
To enable the gradient-checkpointing mode for a graph, we use `setCheckpointing()`:
```

```cpp
// enable gradient-checkpointing for a graph
graph->setCheckpointing(true);
```

> We can also save and load the model (containing all the parameters) in Marian. 

Reviewer: The model is graph + parameters?

Author: The model here only contains the parameters.
